### PR TITLE
Automate generation of JNI bindings.

### LINKED ISSUE #1049: Automate generation of JNI bindings.
### Summary

Now that parser can instantiate scala AST directly in rust, we would like to autogenerate the JNI bindings to scala AST constructors.

### Value

We can get rid of error prone boilerplate.

### Specification

- [x] Create a crate `lib/rust/ast-jni/generation` that generates JNI bindings on modification of rust AST.
- [x] Document how this process works in the design documentation.
- [x] Test the generated code.

### Acceptance Criteria & Test Cases

We can automatically generate the Rust JNI bindings.

### PR DESCRIPTION
**Pull request by Kesanov**
### Pull Request Description

Closes #1049.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: .cargo/config [hunk 1]
```diff
@@ -1,4 +0,0 @@
-
-[build]
-target-dir = "target/rust/"
-rustflags = []
```
Comment (radeusgd) [inline .cargo/config]: Why is this configuration removed?

If the location of built Rust artifacts is now `$rustProjectDir/target/...`, then [the line](https://github.com/enso-org/enso/blob/wip/jv/ast/build.sbt#L814) (and possibly some others?) needs to be updated:

```
    javaOptions += {
      val root = baseDirectory.value.getParentFile.getParentFile.getParentFile
      s"-Djava.library.path=$root/target/rust/debug"
    },
```

Comment (Kesanov) [reply .cargo/config]: Unintended.

File: lib/rust/ast/generation/src/api.rs [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+
+impl Collector {
+    /// Adds a class to collector.
+    pub fn add_class(&mut self, mut class:Class, extends:Option<Name>) {
+        let args = class.args.iter().map(|(name, typ)|
+            (name.clone(), self.monomorphize(&typ).1)
+        ).collect();
+        let arg_types  = class.args.into_iter().map(|(_,typ)| typ);
+        class.typ.path = self.module.clone();
+        let class      = Class{typ:class.typ, args, named:class.named};
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]: This should have a `new` constructor.

File: lib/rust/ast/generation/src/api.rs [hunk 3]
```diff
@@ -0,0 +1,672 @@
...
+        self.classes.push((class.clone(),arg_types.collect(),extends));
+    }
+
+    /// Returns a monomorphized name and the original type with monorphized arguments.
+    pub fn monomorphize(&mut self, typ:&Type) -> (Name, Type) {
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]: Use an actual data type for this return.

File: lib/rust/ast/generation/src/api.rs [hunk 4]
```diff
@@ -0,0 +1,672 @@
...
+        let     path = vec![];
+        for arg in typ.args.iter() {
+            let (name, typ) = self.monomorphize(&arg);
+            uid.push_str(&name.str);
+            args.push(typ);
+        }
+        if args.is_empty() || types::builtin(&typ.name).is_some() {
+            return (Name(uid), Type{name:typ.name.clone(), path, args});
+        }
+        let alias = Type{name:Name(&uid), path, args};
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]: Actual constructor here too please.

File: lib/rust/ast/generation/src/api.rs [hunk 5]
```diff
@@ -0,0 +1,672 @@
...
+    }
+}
+
+
+// === Generator Impls ===
+
+impl Generator<Collector> for &File {
+    fn write(self, source:&mut Collector) {
+        source.package = self.package.clone();
+        Module {name:self.name.clone(), lines:&self.content.items[..]}.write(source);
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]: Constructor please.

File: lib/rust/ast/generation/src/api.rs [hunk 6]
```diff
@@ -0,0 +1,672 @@
+//! This module exports jni bindings generator.
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs:1]: You're missing `new` constructors all throughout this module, and a lot of your doc comments aren't proper sentences.

Additionally, doc comments should be wrriten in an _imperative_ form: `Generate the blahblah`, not `Generates the blahblah`.

File: lib/rust/ast/generation/src/api.rs [hunk 7]
```diff
@@ -0,0 +1,672 @@
...
+    }
+}
+
+
+
+// ======================
+// === Rust Generator ===
+// ======================
+
+/// An associated types in trait with monorphized signature.
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]:
```suggestion
/// An associated type in a trait with a monorphized signature.
```

File: lib/rust/ast/generation/src/api.rs [hunk 8]
```diff
@@ -0,0 +1,672 @@
...
+    /// Name of the associated type.
+    pub name: Name,
+    /// Name of the enum variant if any.
+    pub variant: Option<Name>,
+    /// The class the associated type represents.
+    pub class: Class
+}
+
+impl AssociatedType {
+    /// Ast tree of any type used in trait.
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]:
```suggestion
    /// Ast tree of any type used in a trait.
```

File: lib/rust/ast/generation/src/api.rs [hunk 9]
```diff
@@ -0,0 +1,672 @@
...
+    ///
+    /// For custom types this returns `<Self as Api>::Name`.
+    /// For builtin types this returns `Name<typ(arg1), typ(arg2)..>`.
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]: These sentences should be wrapped properly, not on separate lines.

File: lib/rust/ast/generation/src/api.rs [hunk 10]
```diff
@@ -0,0 +1,672 @@
...
+        if types::builtin(&name).is_none() {
+            quote!(<Self as Api>::#name)
+        } else {
+            quote!(#name<#(#args),*>)
+        }
+    }
+
+    /// An api of function that constructs the given associated type.
+    ///
+    /// For example `fn name(x:i64, y:<Self as Api>::Y) -> <Self as Api>::Name`
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]: This doumentation comment doesn't make any sense. Additionally, if you provide an example you should do it as a proper doctest example in rust.

File: lib/rust/ast/generation/src/api.rs [hunk 11]
```diff
@@ -0,0 +1,672 @@
...
+            let typ = Self::typ(typ);
+            quote!(#name:#typ)
+        });
+
+        quote!(fn #name(&self, #(#args),*) -> <Self as Api>::#typ)
+    }
+
+}
+
+/// A generator of an API for AST construction.
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]:
```suggestion
/// A generator of an API for constructing the AST.
```

File: lib/rust/ast/generation/src/api.rs [hunk 12]
```diff
@@ -0,0 +1,672 @@
...
+    pub fn new(collector:Collector) -> Self {
+        fn apply(typ:&Type, vars:&Map<&Name,&Type>) -> Type {
+            if let Some(&typ) = vars.get(&typ.name) { typ.clone() } else {
+                let args = typ.args.iter().map(|typ| apply(typ,&vars)).collect();
+                Type{name:typ.name.clone(), path:vec![], args}
+            }
+        }
+
+        let mut classes = vec![];
+        let mut types   = vec![];
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]: You should use `default()` here.

File: lib/rust/ast/generation/src/api.rs [hunk 13]
```diff
@@ -0,0 +1,672 @@
...
+            }
+        }
+        Self{class_names:collector.types, classes, package:collector.package, types, generics:collector.generics}
+    }
+
+    /// Generates the AST trait.
+    pub fn ast_trait(&self) -> TokenStream {
+        let types   = self.generics.iter().flat_map(|(_,typ)| typ.iter().map(|typ| &typ.name));
+        let funs    = self.types.iter().map(|obj| obj.fun());
+
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs]: rm nl

File: lib/rust/ast/generation/src/api.rs [hunk 14]
```diff
@@ -0,0 +1,672 @@
...
+// =============
+// === Tests ===
+// =============
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use syn::parse_quote as parse;
+
+
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs:477]:
```suggestion
```

File: lib/rust/ast/generation/src/api.rs [hunk 15]
```diff
@@ -0,0 +1,672 @@
...
+
+    #[test]
+    fn test_monomorphization() {
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs:480]: Loads of extraneous whitespace in the tests. Please remove it.

File: lib/rust/ast/generation/src/ast.rs [hunk 16]
```diff
@@ -0,0 +1,235 @@
+//! This module exports generator ast.
+
+use itertools::Itertools;
+use syn;
+use syn::{Ident, ItemMod};
+use quote::quote;
+use std::ops::Add;
+
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs:10]:
```suggestion


```
Please take better care over your formatting.

File: lib/rust/ast/generation/src/ast.rs [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+    pub name : Name,
+    pub path : Vec<Name>,
+    pub args : Vec<Type>
+}
+
+/// Represents a type alias `type Foo = Bar[Baz]`
+#[allow(missing_docs)]
+#[derive(Debug,Clone)]
+pub struct TypeAlias {
+    pub typ : Type,
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs]: This should be `name`.

File: lib/rust/ast/generation/src/ast.rs [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+        let     generics = &val.generics;
+        let mut variants = vec![];
+        for var in val.variants.iter() {
+            match &var.fields {
+                syn::Fields::Named(_) =>
+                    variants.push((None, Class(Type(&var.ident, generics), &var.fields))),
+                syn::Fields::Unnamed(fields) => {
+                    if let syn::Type::Path(val) = fields.unnamed.first().unwrap().ty.clone() {
+                        let segments             = val.path.segments.iter().rev().take(2);
+                        let (name, object)       = segments.map(|s|Name(&s.ident)).collect_tuple().expect("Unnamed fields in enum must be qualified.");
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs]: This vastly overflows 100 characters.

File: lib/rust/ast/generation/src/ast.rs [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+                _ => (),
+            }
+        }
+        Self{typ:Type(&val.ident, generics), variants}
+    }
+}
+
+impl From<Name> for Type {
+    fn from(name:Name) -> Self {
+        Type{name, path:vec![], args:vec![]}
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs]: Use `default()`.

File: lib/rust/ast/generation/src/ast.rs [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+    }
+}
+
+impl From<&syn::Type> for Type {
+    fn from(typ:&syn::Type) -> Self {
+        let mut name = Name("");
+        let mut path = vec![];
+        let mut args = vec![];
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs]: Use `default()`.

File: lib/rust/ast/generation/src/generator.rs [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+// == Trait Impls ==
+
+impl<S,T:Generator<S>> Generator<S> for Option<T> {
+    fn write(self, source:&mut S) {
+        if let Some(t) = self { t.write(source) }
+    }
+}
+
+
+// == Tab ==
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/generator.rs]: This should be a large heading.

File: lib/rust/ast/generation/src/generator.rs [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+
+/// Writes an amount of spaces corresponding to the current indent.
+#[derive(Debug,Clone,Copy,Default)]
+pub struct Tab();
+
+/// Writes an amount of spaces corresponding to the current indent.
+pub const TAB:Tab = Tab();
+
+
+// == When ==
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/generator.rs]: This should also be a large heading.

File: lib/rust/ast/impl/build.rs [hunk 23]
```diff
@@ -0,0 +1,25 @@
+use std::fs::File;
+use std::io::prelude::*;
+use ast_generation::api;
+
+
+/// Generates the AST API  and saves the result into the file `src/api.rs`.
+///
+/// The content of the generated file can be used with the `include!` macro.
```
Comment (iamrecursion) [inline lib/rust/ast/impl/build.rs]: It shouldn't be used with the `include!` macro. Just generate the file in place instead, like the lexer does.

File: lib/rust/ast/impl/src/ffi.rs [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+
+
+// =====================
+// === JNI Utilities ===
+// =====================
+
+
+// === User Defined Types ===
+
+/// Ergonomic JObject constructor.
```
Comment (iamrecursion) [inline lib/rust/ast/impl/src/ffi.rs]: How can a struct be a constructor?

File: lib/rust/ast/impl/src/ffi.rs [hunk 25]
```diff
@@ -0,0 +1,218 @@
...
+        unimplemented!()
+    }
+}
+
+
+/// Scala standard library.
+mod stdlib {
+    use super::*;
+
+
```
Comment (iamrecursion) [inline lib/rust/ast/impl/src/ffi.rs:203]:
```suggestion
```

File: lib/rust/ast/macros/Cargo.toml [hunk 26]
```diff
@@ -0,0 +1,27 @@
...
+[lib]
+name       = "ast_macros"
+proc-macro = true
+test       = true
+bench      = true
+
+[dependencies]
+quote          = { version = "1.0.7"                          }
+syn            = { version = "1.0.12"                         }
+ast-generation = { version = "0.1.0" , path = "../generation" }
```
Comment (iamrecursion) [inline lib/rust/ast/macros/Cargo.toml]: Missing newline.

File: lib/scala/parser/src/test/scala/org/enso/parser/ParserTest.scala [hunk 27]
```diff
@@ -16,5 +16,6 @@ class ParserTest extends AnyFlatSpec with Matchers {
     assert(expected == parser.parseFile("Hello!"))
     assert(expected == parser.lexStr("Hello!"))
     assert(expected == parser.lexFile("Hello!"))
+    assert(parser.bench().ast.isInstanceOf[Ast.App.Prefix])
```
Comment (iamrecursion) [inline lib/scala/parser/src/test/scala/org/enso/parser/ParserTest.scala]: Not sure why I didn't catch this before. You need to use matchers such as `shouldEqual` and `shouldBe` here. Example:

```scala
parser.bench().ast shouldBe an[Ast.App.Prefix]
```

**Review by iamrecursion — CHANGES_REQUESTED** [review]
A good start, but there are lots of things that need fixing. You need to run clippy on this code.

File: lib/rust/ast/generation/src/api.rs [hunk 28]
```diff
@@ -0,0 +1,706 @@
...
+// ========================
+// === Name Conversions ===
+// ========================
+
+/// Module for name manipulation.
+pub mod name {
+    use crate::ast::Name;
+    use inflector::Inflector;
+
+
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs:448]:
```suggestion
```

File: lib/rust/ast/generation/src/ast.rs [hunk 29]
```diff
@@ -0,0 +1,268 @@
...
+    /// `object FileName { content.. }`, because Scala (compared to Rust) doesn't support
+    /// top level type aliases.
+    pub fn new(name:&str, package:&str, content:syn::File) -> Self {
+        Self{lib:Stdlib(), package:package.into(), name:name.into(), content}
+    }
+}
+
+/// Represents a modified standard library.
+#[derive(Debug,Clone,Copy,Default)]
+pub struct Stdlib();
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs]: What is a modified standard library?

File: lib/rust/ast/generation/src/ast.rs [hunk 30]
```diff
@@ -0,0 +1,268 @@
...
+
+/// Represents a class `class Name[Generics](field:Field, ...)`.
+///
+/// This can either correspond to a Rust struct or a single enum variant.
+#[allow(missing_docs)]
+#[derive(Debug,Clone,Ord,PartialOrd,Eq,PartialEq)]
+pub struct Class {
+    pub typ   : Type,
+    pub args  : Vec<(Name,Type)>,
+    /// True iff the class fields have user defined names.
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs]: If you have a doc you should describe what it is, not just the conditions for setting it.

File: lib/rust/ast/generation/src/ast.rs [hunk 31]
```diff
@@ -0,0 +1,268 @@
...
+    pub named : bool,
+}
+
+/// Represents a set of classes extending a sealed trait.
+#[allow(missing_docs)]
+#[derive(Debug,Clone)]
+pub struct Enum {
+    pub typ      : Type,
+    pub variants : Vec<(Option<Name>, Class)>,
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs]: Don't use tuples.

File: lib/rust/ast/generation/src/ast.rs [hunk 32]
```diff
@@ -0,0 +1,268 @@
...
+/// Class constructor.
+#[allow(non_snake_case)]
+pub fn Class(name:Type, fields:&syn::Fields) -> Class {
+    let named = if let syn::Fields::Named{..} = fields {true} else {false};
+    let args  = fields.iter().enumerate().map(|(i,arg)| {
+        let name = arg.ident.as_ref().map_or(Name(format!("val{}",i)),Name);
+        (name, Type::from(&arg.ty))
+    }).collect();
+
+    Class{typ:name, args, named}
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs]: You need to use the `::new` functions that you've written.

File: lib/rust/ast/generation/src/ast.rs [hunk 33]
```diff
@@ -0,0 +1,268 @@
...
+    fn from(val:&syn::ItemStruct) -> Self {
+        Class(Type(&val.ident, &val.generics), &val.fields)
+    }
+}
+
+
+impl<'a> From<&'a syn::ItemMod> for Module<'a> {
+    fn from(val:&'a ItemMod) -> Self {
+        let lines = if let Some((_,lines)) = &val.content {&lines[..]} else {&[]};
+
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs]:
```suggestion
```

File: lib/rust/ast/generation/src/generator.rs [hunk 34]
```diff
@@ -0,0 +1,58 @@
...
+        source
+    }
+}
+
+
+// == Trait Impls ==
+
+impl<S,T:Generator<S>> Generator<S> for Option<T> {
+    fn write(self, source:&mut S) {
+        if let Some(t) = self { t.write(source) }
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/generator.rs]:
```suggestion
        if let Some(t) = self{t.write(source)}
```

File: lib/rust/ast/generation/src/scala.rs [hunk 35]
```diff
@@ -0,0 +1,283 @@
...
+
+    /// Genereate a scala source code.
+    pub fn ast() -> std::io::Result<String> {
+        let mut content = String::new();
+        let mut file    = fs::File::open("lib/rust/ast/impl/src/ast.rs")?;
+        file.read_to_string(&mut content)?;
+
+        let file      = syn::parse_file(content.as_str()).unwrap();
+        let this:Self = File::new("Ast", "org.enso.ast", file).source();
+
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/scala.rs]: Blank lines?

File: lib/rust/ast/generation/src/scala.rs [hunk 36]
```diff
@@ -0,0 +1,283 @@
...
+impl<'a> Generator<Source> for &Module<'a> {
+    fn write(self, source:&mut Source) {
+        let name = name::typ(&self.name);
+
+        write!(source, TAB, "object ", &name, " {\n");
+
+        source.indent += 2;
+
+        if source.extends.contains_key(&name) {
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/scala.rs]: Blank lines?

File: lib/rust/ast/generation/src/scala.rs [hunk 37]
```diff
@@ -0,0 +1,283 @@
...
+                syn::Item::Type  (val) => { source.write("\n"); TypeAlias::from(val).write(source) }
+                syn::Item::Struct(val) => { source.write("\n"); Class    ::from(val).write(source) }
+                syn::Item::Enum  (val) => { source.write("\n"); Enum     ::from(val).write(source) }
+                _                      => { }
+            }
+        }
+
+        source.indent -= 2;
+
+        write!(source, TAB, "}\n");
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/scala.rs]: Here too.

File: lib/rust/ast/generation/src/scala.rs [hunk 38]
```diff
@@ -0,0 +1,283 @@
+//! This module exports scala ast generator.
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/scala.rs:1]: You have loads of blank lines in this file. Please get rid of them.

File: lib/rust/ast/impl/build.rs [hunk 39]
```diff
@@ -0,0 +1,39 @@
...
+    let ast_error    = format!("The AST definition should exist at {}.", ast_path);
+    let api_error    = format!("Cannot open API file at {}.", api_path);
+    let mut ast_file = File::open(ast_path).expect(ast_error.as_str());
+    let mut api_file = File::create(api_path).expect(api_error.as_str());
+    let mut ast_str  = String::new();
+
+    ast_file.read_to_string(&mut ast_str).expect("Unable to read AST definition.");
+    let api = api::Source::api(ast_str.as_str());
+    api_file.write_all(api.as_bytes()).expect("Unable to write API definition.");
+
```
Comment (iamrecursion) [inline lib/rust/ast/impl/build.rs]: Blank lines.

File: lib/rust/ast/impl/src/ffi.rs [hunk 40]
```diff
@@ -0,0 +1,258 @@
...
+    ///
+    /// Example:
+    /// ```rust,ignore
+    /// Object::new(env, "org/enso/ast/Ast$Ast", "(Lscala/Option;JJLjava/lang/Object;)V")
+    /// ```
+    pub fn new(env:&'a JNIEnv<'a>, typ:&str, args:&str) -> Self {
+        let err = "Could not find class ".to_string() + typ;
+        let obj = env.find_class(typ).expect(&err);
+        let fun = env.get_method_id(typ, "<init>", args).expect(&(err+" method "+args));
+
```
Comment (iamrecursion) [inline lib/rust/ast/impl/src/ffi.rs]: Blank line.

File: project/GenerateAST.scala [hunk 41]
```diff
@@ -9,17 +9,17 @@ object GenerateAST {
...
 
     Tracked.diffInputs(cache, FileInfo.lastModified)(Set(source))
     { source: ChangeReport[File] =>
       val rustVersion = Cargo.rustVersion.value
-      if (source.modified.nonEmpty) {
+//      if (source.modified.nonEmpty) {
           output.getParentFile.mkdirs
           generateAST(rustVersion, output, log)
-      }
+//      }
```
Comment (iamrecursion) [inline project/GenerateAST.scala]: Why is this commented out?

**Review by iamrecursion — CHANGES_REQUESTED** [review]
Please make sure that you complete the documentation, including the performance analysis, and _carefully_ check all of your code for style. There are so many deviations from the style guide. I haven't pointed them all out.

File: docs/parser/ast-generation.md [hunk 42]
```diff
@@ -0,0 +1,234 @@
...
+
+Note that builtin types (types that are not defined in `ast.rs`) are left intact 
+(they do not introduce an additional associated type). Therefore, they need
+to implement the trait `ast::ffi::ToJValue` to enable the autogenerated
+[Scala Bindings](#scala-bindings) to convert those types to correct JNI values.
+
+## Performance
+
+On benchmarks that allocate 2e6 AST nodes the Rust API takes ~0.4s while the 
+Scala API takes ~4s.
```
Comment (iamrecursion) [inline docs/parser/ast-generation.md:230]: You need to describe the benchmarking and profiling approaches taken here.

Additionally, you should test instantiation of something small in a loop via JNI and check the overhead. I would test via `jni-rs`, but also via the c bindings directly.

File: lib/rust/ast/generation/src/api.rs [hunk 43]
```diff
@@ -12,82 +12,98 @@ use quote::quote;
...
+    pub args    : Vec<Type>,
+    pub extends : Option<Name>,
+}
+
+impl MonomorphizedClass {
+    /// Construct a new instance.
+    pub fn new(class:Class, args:Vec<Type>, extends:Option<Name>) -> Self {
+        Self{class,args,extends}
+    }
+}
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs:34]: This should be in its own section.

File: lib/rust/ast/generation/src/api.rs [hunk 44]
```diff
@@ -12,82 +12,98 @@ use quote::quote;
...
+        UniqueType::new(Name(&uid),Type::from(Name(&uid)))
     }
 }
 
 /// Representation of type with an additional unique name.
 #[derive(Debug,Clone)]
 pub struct UniqueType {
     /// Unique name of this type.
-    uid: Name,
+    unique_name: Name,
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs:88]:
```suggestion
    unique_name : Name,
```

File: lib/rust/ast/generation/src/api.rs [hunk 45]
```diff
@@ -229,55 +245,63 @@ impl Source {
...
     }
 
     /// Create a new instance from `Collector`.
-    pub fn new(collector: PreprocessedFile) -> Self {
+    pub fn new(file:PreprocessedFile) -> Self {
         fn apply(typ:&Type, vars:&Map<&Name,&Type>) -> Type {
             if let Some(&typ) = vars.get(&typ.name) { typ.clone() } else {
                 let args = typ.args.iter().map(|typ| apply(typ,&vars)).collect();
-                Type::new(typ.name.clone(), vec![], args)
+                Type::new(typ.name.clone(),Vec::default(),args)
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs:258]: This should be inferred, so you should be able to use bare `default()`. Specifying the type you're defaulting kinda removes the usefulness of using default over `vec![]`.

File: lib/rust/ast/generation/src/api.rs [hunk 46]
```diff
@@ -454,43 +478,49 @@ mod tests {
...
             struct A<T>(T);
+            /// Wrapper around A<i32>.
             struct B(A<i32>);
         }).into());
         let expected = quote! {
             pub trait Api {
+                #[doc="The AST Node A<i32>."]
                 type Ai32;
+                #[doc="The AST Node B."]
                 type B;
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs:511]:
```suggestion
```

File: lib/rust/ast/generation/src/api.rs [hunk 47]
```diff
@@ -512,27 +542,27 @@ mod tests {
...
             struct B<X,Y>(X,Y);
             struct C<T>(T);
         }).into());
         let expected = quote! {
             impl Api for Rust {
                 type A         = A<>;
-                type BXBoxi32  = B< <Self as Api>::X, Box<i32 <> > >;
+                type BXBoxi32  = B< <Self as Api>::X, Box<i32> >;
                 type BXY       = B< <Self as Api>::X, <Self as Api>::Y>;
                 type CBXBoxi32 = C< <Self as Api>::BXBoxi32>;
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/api.rs:561]:
```suggestion
```
Please remove unneeded blank lines.

File: lib/rust/ast/generation/src/ast.rs [hunk 48]
```diff
@@ -52,18 +47,35 @@ pub struct Module<'a> {
 #[allow(missing_docs)]
 #[derive(Debug,Clone,Ord,PartialOrd,Eq,PartialEq)]
 pub struct Class {
+    pub doc   : String,
     pub typ   : Type,
-    pub args  : Vec<(Name,Type)>,
-    /// True iff the class fields have user defined names.
+    pub args  : Vec<Field>,
+    /// The class fields have user defined names.
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs:53]:
```suggestion
    /// The class fields have user defined names iff this is true.
```

File: lib/rust/ast/generation/src/ast.rs [hunk 49]
```diff
@@ -249,8 +273,8 @@ impl From<&str> for Type {
 impl From<&syn::Type> for Type {
     fn from(typ:&syn::Type) -> Self {
         let mut name = Name::default();
-        let mut path = Vec::default();
-        let mut args = Vec::default();
+        let mut path = Vec ::default();
+        let mut args = Vec ::default();
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/ast.rs:277]: Don't put a space after `Vec`, but you also shouldn't be specifying what you're defaulting.

File: lib/rust/ast/generation/src/types.rs [hunk 50]
```diff
@@ -56,30 +57,30 @@ pub struct JName {
...
-       "usize" | "isize" | "u64" | "i64" => JName {jni:"J", scala:"Long"   },
-       "u32"   | "i32"                   => JName {jni:"I", scala:"Int"    },
-       "u16"   | "i16"   | "i8"          => JName {jni:"S", scala:"Short"  },
-       "u8"                              => JName {jni:"B", scala:"Byte"   },
-       "f64"                             => JName {jni:"D", scala:"Double" },
-       "f32"                             => JName {jni:"F", scala:"Float"  },
-       "boolean"                         => JName {jni:"Z", scala:"Boolean"},
-       "char"                            => JName {jni:"C", scala:"Char"   },
-        _                                => stdlib(name)?,
+       "i64"          => JName {jni:"J", scala:"Long"   },
```
Comment (iamrecursion) [inline lib/rust/ast/generation/src/types.rs:62]: Why did you remove the rest of these?

**Review by iamrecursion — CHANGES_REQUESTED** [review]

File: lib/rust/ast/generation/src/ast.rs (hunk 49 above)
Comment (Kesanov) [reply lib/rust/ast/generation/src/ast.rs:277]: It cannot infer type.

File: lib/rust/ast/generation/src/types.rs (hunk 50 above)
Comment (Kesanov) [reply lib/rust/ast/generation/src/types.rs:62]: Because they can overflow. I will add `isize` and `u32` back though.

**Comment by iamrecursion** [comment]
Closing this until the work gets picked up again.
